atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,290 @@
 of the BSD license. See the LICENSE file for details.
 """
 
-import os
-import logging
-import json
-import paramiko
 import backoff
+import datetime
+import json
 import jsonschema
-from typing import List, Any, Dict, Tuple
+import paramiko
+import random
+import os
 from datetime import datetime
 from shlex import quote
+from typing import Any, Dict, List, Tuple
+
+SSH_COMMAND_TIMEOUT = 120
+RELATIVE_PATH = 'resource_manager'
+RESOURCE_INFO_JSON = 'info.json'
+RETRY_ON_SSH_EXCEPTIONS = (paramiko.ssh_exception.AuthenticationException,
+                           paramiko.ssh_exception.ChannelException,
+                           paramiko.ssh_exception.NoValidConnectionsError,
+                           paramiko.ssh_exception.SSHException)
+SSH_BACKOFF_FACTOR = 5
+SSH_MAX_RETRIES = 3
 
-PARAMIKO_MAX_RETRIES = 3
-PARAMIKO_BACKOFF_FACTOR = 1
 
-ROOT_PATH = 'resource_manager/'
+class LockError(Exception):
+    """ Attempt for resource locking returned non-zero exitcode """
+    pass
 
-RESOURCE_LOCK_SCHEMA = 'schemas/resource-lock-content.json'
 
-logger = logging.getLogger(__name__)
+class LockedResource:
 
+    def __init__(host: RemoteHost, slot: int, prid: str):
+        """
+        :param slot: int, Remote host slot ID
+        :param prid: str, Pipeline run ID
+        """
+        self.slot = slot
+        self.prid = prid
 
-class Host:
-    '''
-    Remote host abstract class
-    '''
-    def __init__(self, username: str, hostname: str):
-        self.username = username
-        self.hostname = hostname
-        self.slots: List[Dict[str, Any]] = []
-        self.free_slots: List[int] = []
+    def unlock(self):
+        self.host.unlock(self.slot, self.prid)
 
-    def _read_info_data(self):
-        cmd = f"cat {ROOT_PATH}info.json"
-        data = self._ssh_run_remote_command(cmd)
-        return data
 
-    def _read_slot_data(self, slot: int):
-        cmd = f"cat {ROOT_PATH}slot_{slot}.json"
-        data = self._ssh_run_remote_command(cmd)
-        return data
+class RemoteHost:
+
+    def __init__(self, hostname: str, username: str, ssh_keyfile: str):
```

Comment:
There will be more attrs, so it's better to enforce params by name
```suggestion
    def __init__(self, *, hostname: str, username: str, ssh_keyfile: str):
```

Reply:
I agree